In [1]:
import pandas as pd

import warnings
warnings.filterwarnings("ignore")

In [2]:
df = pd.read_csv("spam.csv")

df.head()

Category                                            Message
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...

In [3]:
df.groupby(df.Category).describe()

Message                                                            \
           count unique                                                top   
Category                                                                     
ham         4825   4516                             Sorry, I'll call later   
spam         747    641  Please call our customer service representativ...   

               
         freq  
Category       
ham        30  
spam        4

### DATA PREPROCESSING

There is a total of 747 spam messages

In [4]:
# First lets convert spam -> 1, ham -> 0
df['Category'] = df['Category'].apply(lambda x: 1 if x=='spam' else 0)
df.head()

Category                                            Message
0         0  Go until jurong point, crazy.. Available only ...
1         0                      Ok lar... Joking wif u oni...
2         1  Free entry in 2 a wkly comp to win FA Cup fina...
3         0  U dun say so early hor... U c already then say...
4         0  Nah I don't think he goes to usf, he lives aro...

### TRAINING

In [5]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(df.Message, df.Category, test_size=0.3)

In [ ]:
# from sklearn.feature_extraction.text import CountVectorizer

# cv = CountVectorizer()

In [ ]:
# we will transform x train values using CountVector and convert it to array
# X_cv = cv.fit_transform(X_train.values)
# X_cv.toarray()

Now we have **"X_cv"** and **"y_train"** to train our model

In [ ]:
# from sklearn.naive_bayes import MultinomialNB

# mnb = MultinomialNB()

In [ ]:
# mnb.fit(X_cv, y_train)

### Testing our model
first we need to transform **'X_test'** using CountVector

In [ ]:
# X_test_cv = cv.transform(X_test)
# X_test_cv.toarray()

In [ ]:
# mnb.score(X_test_cv, y_test)

In [ ]:
# emails = [
#     'Hey mohan, can we get together to watch footbal game tomorrow?',
#     'Upto 20% discount on parking, exclusive offer just for you. Dont miss this reward!'
# ]
# email_count = cv.transform(emails)
# mnb.predict(email_count)

### Another method is to use Sklearn Pipelines

In [6]:
from sklearn.naive_bayes import BernoulliNB
from sklearn.naive_bayes import MultinomialNB
from sklearn.naive_bayes import GaussianNB

from sklearn.feature_extraction.text import CountVectorizer

In [16]:
from sklearn.pipeline import Pipeline
model1 = Pipeline([
    ('cv1', CountVectorizer()),
    ('bnb', BernoulliNB())
])
model2 = Pipeline([
    ('cv2', CountVectorizer()),
    ('mnb', MultinomialNB())
])

In [17]:
model1.fit(X_train, y_train)

Pipeline(steps=[('cv1', CountVectorizer()), ('bnb', BernoulliNB())])

In [18]:
model1.score(X_test, y_test)

0.9706937799043063

In [19]:
model2.fit(X_train, y_train)

Pipeline(steps=[('cv2', CountVectorizer()), ('mnb', MultinomialNB())])

In [20]:
model2.score(X_test, y_test)

0.9838516746411483